src/liballoc/macros.rs

```diff
@@ -39,9 +39,20 @@ macro_rules! vec {
     ($elem:expr; $n:expr) => (
         $crate::vec::from_elem($elem, $n)
     );
-    ($($x:expr),*) => (
-        <[_]>::into_vec(box [$($x),*])
-    );
+    ($($x:expr),*) => ({
+        // We use a temporary let binding to work around a coercion
+        // bug/strangness with `box [a, b]` syntax. Example:
+        //
+        //     fn foo() {}
+        //     fn bar() {}
+        //
+        //     let _ = [foo, bar]; // works: fn items coerced to fn pointers
+        //     let _ = <[_]>::into_vec(box [foo, bar]);  // doesn't work
+        //
+        // See the `macro_fn_pointer_coercion` test.
+        let tmp = [$($x),*];
+        <[_]>::into_vec(box tmp)
```

Comment:
Possibly-silly question: now that there's a temporary, is `box` syntax gaining us anything over using `Box::new`?  Is there a chance that some code that was making too-big-for-the-stack `Vec`s with this will no longer work?

Reply:
Alternatively, is this something that should be fixed in the compiler about `box` syntax?  Or is that dead enough that we should just stop ever using the keyword in the library?

Reply:
Good point. I guess people could use `vec!` with data too big for the stack, so we shouldn't change this as it might break code. I don't think there is a way to easily fix that, so I will close this PR.

The correct way would be to fix `box`, I guess. But I heard it's not gonna land in this form anyway, so we might also just ignore this obscure function-pointer problem for now.